fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/SynchronizedSessionTest.java

```diff
@@ -0,0 +1,325 @@
+            checkActive(session1Runner1);
+
+            session1Runner1.closeSession();
+
+            // Runners of the current session should not be able to work (neither existing runner nor newly created runner).
+            assertFailedContinueSession(session1Runner1);
+            try (SynchronizedSessionRunner session1Runner2 = newRunnerJoinSession(lockSubspace, session1Runner1.getSessionId())) {
+                assertFailedContinueSession(session1Runner2);
+            }
+
+            // The new session should be able to be created and used right away.
+            try (SynchronizedSessionRunner session2Runner = newRunnerStartSession(lockSubspace)) {
+                checkActive(session2Runner);
+            }
+        }
+    }
+
+    @Test
+    public void takeLaterOneWhenThereAreDifferentLeaseEndTimes() throws Exception {
+        final Subspace lockSubspace = new Subspace(Tuple.from(Math.random()));
+        try (SynchronizedSessionRunner session1Runner1 = database.newRunner().toSynchronized(lockSubspace, 2_000)) {
+            checkActive(session1Runner1);
+
+            UUID session1Id = session1Runner1.getSessionId();
+            try (SynchronizedSessionRunner session1Runner2 = database.newRunner().toSynchronized(lockSubspace, session1Id, 4_000)) {
+                checkActive(session1Runner2);
+                checkActive(session1Runner1);
+
+                AtomicBoolean start = new AtomicBoolean(false);
+                CompletableFuture<Void> session1Runner1Job = session1Runner1.runAsync(context -> AsyncUtil.whileTrue(
+                        () -> CompletableFuture.completedFuture(start.get()),
+                        context.getExecutor()));
+                CompletableFuture<Void> session1Runner2Job = session1Runner1.runAsync(context -> AsyncUtil.whileTrue(
+                        () -> CompletableFuture.completedFuture(start.get()),
+                        context.getExecutor()));
+                database.newRunner().asyncToSync(null, CompletableFuture.allOf(session1Runner1Job, session1Runner2Job));
+
+                // Make the the jobs run almost at the same time with unknown order.
+                start.set(true);
+
+                // Runner 1 set the lease end time to 2 seconds in the future, while Runner 2 set it to 4 seconds,
+                // the later one should be honoured. So a new session shouldn't take the lock until 4 seconds.
+                Thread.sleep(2_000 + 100);
+                assertFailedStartSession(lockSubspace);
+
+                Thread.sleep(2_000);
+                try (SynchronizedSessionRunner session2Runner = newRunnerStartSession(lockSubspace)) {
+                    checkActive(session2Runner);
+                }
+            }
+        }
+    }
+
+    @Test
+    public void singleRunnerRenewLeaseContinuously() throws Exception {
+        testRenewLeaseContinuously(true);
+    }
+
+    @Test
+    public void newRunnersRenewLeaseContinuously() throws Exception {
```

Comment:
Could these two tests get marked as `Slow` (with `@Tag(Tags.Slow)`)?

Reply:
All of the slow tests run in CI. One can disable the slow tests locally, though, by adding the, um, `SKIP_SLOW_TESTS` environment variable, though I've also seen that not work.

This is just a courtesy to those people so that they can enspeeden their builds.